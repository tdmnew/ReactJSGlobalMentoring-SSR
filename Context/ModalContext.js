import React from "react";

const useModalState = (initialState) => {
    const [isOpen, setIsOpen] = React.useState(initialState.isOpen);
    const [modalProps, setModalProps] = React.useState(initialState.modalProps);

    const setModalState = ({ isOpen, modalProps = {} }) => {
        setIsOpen(isOpen);
        setModalProps(modalProps);
    };

    return [{ isOpen, modalProps }, setModalState];
};

export const ModalStateContext = React.createContext();
export const ModalUpdaterContext = React.createContext();

export const ModalContext = ({ children }) => {
    const [modalOptions, setModalOptions] = useModalState({
        isOpen: false,
        modalProps: {},
    });

    return (
            <ModalUpdaterContext.Provider value={setModalOptions}>
                <ModalStateContext.Provider value={modalOptions}>
                    {children}
                </ModalStateContext.Provider>
            </ModalUpdaterContext.Provider>
    );
};
