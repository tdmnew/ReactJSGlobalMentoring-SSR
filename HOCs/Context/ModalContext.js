import React, { createContext, useState } from 'react';

const useModalState = (initialState) => {
  const [isOpen, setIsOpen] = useState(initialState.isOpen);
  const [modalProps, setModalProps] = useState(initialState.modalProps);

  const setModalState = ({ isOpen, modalProps = {} }) => {
    setIsOpen(isOpen);
    setModalProps(modalProps);
  };

  return [{ isOpen, modalProps }, setModalState];
};

export const ModalStateContext = createContext();
export const ModalUpdaterContext = createContext();

export const ModalContext = ({ children }) => {
  const [modalOptions, setModalOptions] = useModalState({
    isOpen: false,
    modalProps: {},
  });

  return (
    <ModalUpdaterContext.Provider value={setModalOptions}>
      <ModalStateContext.Provider value={modalOptions}>
        {children}
      </ModalStateContext.Provider>
    </ModalUpdaterContext.Provider>
  );
};
