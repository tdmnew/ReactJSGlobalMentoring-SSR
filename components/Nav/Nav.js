import React from "react";

//Redux
import { useDispatch, useSelector } from "react-redux";
import { sagaActions } from "../../redux/sagas/sagaActions";
import { sortMovies } from "../../redux/slices/movies.js";

export default function Nav() {
    const dispatch = useDispatch();
    const movies = useSelector((state) => state.movies);
    const [genres, setGenres] = React.useState([]);

    React.useEffect(() => {
        if (movies !== undefined) {
            const genres = movies
                .map((movie) => movie.genres)
                .flat()
                .filter((v, i, s) => s.indexOf(v) === i);
            setGenres(genres);
        }
    }, [movies]);

    const sort = (e) => {
        dispatch(sortMovies(e.target.value));
    };

    const filter = (e) => {
        const genre = e.target.getAttribute("href").slice(1);
        if (genre === "All") {
            dispatch({ type: sagaActions.FETCH_MOVIES });
        } else {
            dispatch({ type: sagaActions.FILTER_MOVIES, payload: genre });
        }
    };

    return (
        <div className="nav">
            <ul className="nav__links">
                <li>
                    <a href="#All" onClick={filter}>
                        ALL
                    </a>
                </li>
                {genres.slice(0, 6).map((genre) => {
                    return (
                        <li key={genre}>
                            <a href={`#${genre}`} onClick={filter}>
                                {genre}
                            </a>
                        </li>
                    );
                })}
            </ul>
            <div className="dropdown">
                <span className="dropdown__heading">SORT BY</span>
                <select className="dropdown__options" onChange={sort}>
                    <option value="release_date">RELEASE DATE</option>
                    <option value="title">TITLE</option>
                    <option value="genres">GENRE</option>
                    <option value="vote_average">RATING</option>
                </select>
            </div>
        </div>
    );
}
