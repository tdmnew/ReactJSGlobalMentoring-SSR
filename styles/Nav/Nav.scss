.nav {
    font-weight: lighter;
    letter-spacing: 1px;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    border-bottom: 2px solid grey;
    color: #ffffff;
    background: #232323;
    width: 46rem;
    font-size: 12px;

    @media (max-width: 815px) {
        width: auto;
    }

    @media (max-width: 570px) {
        font-size: 9px;
    }

    @media (max-width: 460px) {
        flex-wrap: wrap;
        justify-content: center;
    }

    &__links {
        display: flex;
        list-style-type: none;
        margin: 1rem 0;
        padding: 0;

        @media (max-width: 450px) {
            flex-wrap: wrap;
            justify-content: center;
        }

        li {
            display: inline;
            padding-right: 1rem;

            @media (max-width: 450px) {
                padding: 0 0.5rem;
            }
        }

        .link {
            color: inherit;
            text-decoration: none;
            padding: 1rem 0;
            
            @media (max-width: 570px) {
                padding: .15rem 0;
            }

            &--active {
                @extend .link;
                border-bottom: 2px solid #f65261;
            }
        }
    }

    .dropdown {
        @media (max-width: 815px) {
            padding-left: 1rem;
        }

        @media (max-width: 450px) {
            padding: 0.5rem;
        }

        &__heading {
            color: grey;
            padding-right: 1rem;
            font-weight: bold;

            @media (max-width: 815px) {
                padding-right: inherit;
            }
        }

        &__options {
            color: white;
            font-size: 12px;
            font-weight: lighter;
            background: none;
            border: none;

            option { 
                background: #232323;
            }

            @media (max-width: 570px) {
                font-size: 9px;
            }
        }

    }
}
