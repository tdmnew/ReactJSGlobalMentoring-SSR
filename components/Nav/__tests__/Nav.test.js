import React from "react";

import { getByDisplayValue, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";

import Nav from "../Nav.js";
import { ModalStateContext } from "../../../Context/ModalContext.js";

describe("Nav", () => {
    const initialState = {
        movies: [],
        selectedMovie: {}
    }

    const mockStore = configureStore();
    const store = mockStore(initialState);

    it("Renders the nav bar and default properties", () => {
        const modalOptions = {
            isOpen: false,
        };

        const tree = render(
            <Provider store={store}>
                <ModalStateContext.Provider value={modalOptions}>
                    <Nav />
                </ModalStateContext.Provider>
            </Provider>
        );

        expect(tree).toMatchSnapshot();

        expect(screen.getByText(/ALL/i)).toBeInTheDocument();
        expect(screen.getByText(/RELEASE DATE/i)).toBeInTheDocument();
        expect(screen.getByText(/TITLE/i)).toBeInTheDocument();
        expect(screen.getByText(/GENRE/i)).toBeInTheDocument();
        expect(screen.getByText(/RATING/i)).toBeInTheDocument();
    });
});
