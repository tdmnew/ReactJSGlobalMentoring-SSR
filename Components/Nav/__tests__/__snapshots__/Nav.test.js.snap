// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav Renders the nav bar and default properties 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <i18nextprovider
        i18n="[object Object]"
      >
        <div
          class="nav"
        >
          <ul
            class="nav__links"
          >
            <li>
              <a
                class="link"
                href="#ALL"
              >
                ALL
              </a>
            </li>
          </ul>
          <div
            class="dropdown"
          >
            <span
              class="dropdown__heading"
            >
              SORT BY
            </span>
            <select
              class="dropdown__options"
            >
              <option
                value="release_date"
              >
                RELEASE DATE
              </option>
              <option
                value="title"
              >
                TITLE
              </option>
              <option
                value="genres"
              >
                GENRE
              </option>
              <option
                value="vote_average"
              >
                RATING
              </option>
            </select>
          </div>
        </div>
      </i18nextprovider>
    </div>
  </body>,
  "container": <div>
    <i18nextprovider
      i18n="[object Object]"
    >
      <div
        class="nav"
      >
        <ul
          class="nav__links"
        >
          <li>
            <a
              class="link"
              href="#ALL"
            >
              ALL
            </a>
          </li>
        </ul>
        <div
          class="dropdown"
        >
          <span
            class="dropdown__heading"
          >
            SORT BY
          </span>
          <select
            class="dropdown__options"
          >
            <option
              value="release_date"
            >
              RELEASE DATE
            </option>
            <option
              value="title"
            >
              TITLE
            </option>
            <option
              value="genres"
            >
              GENRE
            </option>
            <option
              value="vote_average"
            >
              RATING
            </option>
          </select>
        </div>
      </div>
    </i18nextprovider>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
